x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForActiveShardsStep.java

```diff
@@ -64,43 +67,50 @@ public Result isConditionMet(Index index, ClusterState clusterState) {
         final String waitForActiveShardsSettingValue;
-        if (aliasWriteIndex != null) {
-            rolledIndexName = aliasWriteIndex.getIndex().getName();
-            waitForActiveShardsSettingValue = aliasWriteIndex.getSettings().get("index.write.wait_for_active_shards");
-        } else {
-            List<IndexMetadata> indices = indexAbstraction.getIndices();
-            int maxIndexCounter = -1;
-            IndexMetadata rolledIndexMeta = null;
-            for (IndexMetadata indexMetadata : indices) {
-                int indexNameCounter = parseIndexNameCounter(indexMetadata.getIndex().getName());
-                if (maxIndexCounter < indexNameCounter) {
-                    maxIndexCounter = indexNameCounter;
-                    rolledIndexMeta = indexMetadata;
-                }
-            }
+        if (indexAbstraction.getParentDataStream() != null) {
+            DataStream dataStream = indexAbstraction.getParentDataStream().getDataStream();
+            IndexAbstraction dataStreamAbstraction = metadata.getIndicesLookup().get(dataStream.getName());
+            assert dataStreamAbstraction != null : dataStream.getName() + " datastream is not present in the metadata indices lookup";
+            IndexMetadata rolledIndexMeta = dataStreamAbstraction.getWriteIndex();
             if (rolledIndexMeta == null) {
-                String errorMessage = String.format(Locale.ROOT,
-                    "unable to find the index that was rolled over from [%s] as part of lifecycle action [%s]", index.getName(),
-                    getKey().getAction());
-
-                // Index must have been since deleted
-                logger.debug(errorMessage);
-                return new Result(false, new Info(errorMessage));
+                return getErrorResultOnNullMetadata(index);
             }
             rolledIndexName = rolledIndexMeta.getIndex().getName();
             waitForActiveShardsSettingValue = rolledIndexMeta.getSettings().get("index.write.wait_for_active_shards");
+        } else {
+            String rolloverAlias = RolloverAction.LIFECYCLE_ROLLOVER_ALIAS_SETTING.get(originalIndexMeta.getSettings());
+            if (Strings.isNullOrEmpty(rolloverAlias)) {
+                throw new IllegalStateException("setting [" + RolloverAction.LIFECYCLE_ROLLOVER_ALIAS
+                    + "] is not set on index [" + originalIndexMeta.getIndex().getName() + "]");
+            }
+
+            IndexAbstraction aliasAbstraction = metadata.getIndicesLookup().get(rolloverAlias);
+            assert aliasAbstraction.getType() == IndexAbstraction.Type.ALIAS : rolloverAlias + " must be an alias but it is not";
+
+            IndexMetadata aliasWriteIndex = aliasAbstraction.getWriteIndex();
+            if (aliasWriteIndex != null) {
+                rolledIndexName = aliasWriteIndex.getIndex().getName();
+                waitForActiveShardsSettingValue = aliasWriteIndex.getSettings().get("index.write.wait_for_active_shards");
```

Comment:
Super minor, but can we use `IndexMetadata.SETTING_WAIT_FOR_ACTIVE_SHARDS` instead of hardcoding here?